keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,456 @@
+
+### Non-Goals
+
+- Provide a way to run the kubelet process or container runtimes as an
+  unprivileged process. Although initiatives like [kubelet in user
+  namespaces][kubelet-userns] and this KEP both make use of user namespaces, it is
+  a different implementation for a different purpose.
+- Implement all the very nice use cases that user namespaces allows. The goal
+  here is to allow them as incremental improvements, not implement all the
+  possible ideas related with user namespaces.
+
+[kubelet-userns]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2033-kubelet-in-userns-aka-rootless
+
+## Proposal
+
+This proposal aims to support running pods inside user namespaces. This will
+improve the pod to node isolation (phase 1 and 2) and pod to pod isolation
+(phase 3) we currently have.
+
+This mitigates all the vulnerabilities listed in the motivation section.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to grant me any extra privileges on the host or other pods.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to run in the host user namespace
+if they need a feature only available in such user namespace, such as loading a
+kernel module with `CAP_SYS_MODULE`.
+
+#### Story 3
+
+As a cluster admin, I want to allow users to run their container as root
+without that process having root privileged on the host, so I can mitigate the
+impact of a compromised container.
+
+#### Story 4
+
+As a cluster admin, I want to allow users to choose the user/group they use
+inside the container without it being the user/group they use on the node, so I
+can mitigate the impact of a container breakout vulnerability (e.g to read
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running in
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+Note: Names are preliminary yet, I'm using field names to simplify explanations.
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.useHostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
+If false, a new userns is created for the pod.
+This field will be used for phase 1, 2 and 3.
+
+- `pod.spec.securityContext.userns.pod2podIsolation`: Enum
+If enabled, we will make the userns mappings be non-overlapping as much as possible.
+This field will be used in phase 3.
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
```

Comment:
Could you clarify if each phase would be a separate KEP?

Reply:
The initial idea was at least phase 1 and phase 2. Phase 3 says it will be a different KEP if it turns out too complex.